agent/exec/containerd/controller.go

```diff
@@ -0,0 +1,401 @@
+	adapter *containerAdapter
+	closed  chan struct{}
+	err     error
+
+	pulled     chan struct{} // closed after pull
+	cancelPull func()        // cancels pull context if not nil
+	pullErr    error         // pull error, protected by close of pulled
+}
+
+var _ exec.Controller = &controller{}
+
+func newController(client execution.ExecutionServiceClient, bundleDir, contentStore string, task *api.Task, secrets exec.SecretGetter) (exec.Controller, error) {
+	adapter, err := newContainerAdapter(client, bundleDir, contentStore, task, secrets)
+	if err != nil {
+		return nil, err
+	}
+
+	return &controller{
+		task:    task,
+		adapter: adapter,
+		closed:  make(chan struct{}),
+	}, nil
+}
+
+// Update tasks a recent task update and applies it to the container.
+func (r *controller) Update(ctx context.Context, t *api.Task) error {
+	log.G(ctx).Warnf("task updates not yet supported")
+	// TODO(stevvooe): While assignment of tasks is idempotent, we do allow
+	// updates of metadata, such as labelling, as well as any other properties
+	// that make sense.
+	return nil
+}
+
+// Prepare creates a container and ensures the image is pulled.
+//
+// If the container has already be created, exec.ErrTaskPrepared is returned.
+// XXX currently does none of this
+func (r *controller) Prepare(ctx context.Context) error {
+	if err := r.checkClosed(); err != nil {
+		return err
+	}
+
+	//// Make sure all the networks that the task needs are created.
+	//if err := r.adapter.createNetworks(ctx); err != nil {
+	//	return err
+	//}
+
+	//// Make sure all the volumes that the task needs are created.
+	//if err := r.adapter.createVolumes(ctx); err != nil {
+	//	return err
+	//}
+
+	if r.pulled == nil {
+		// Launches a re-entrant pull operation associated with controller,
+		// dissociating the context from the caller's context. Allows pull
+		// operation to be re-entrant on calls to prepare, resuming from the
+		// same point after cancellation.
+		var pctx context.Context
+
+		r.pulled = make(chan struct{})
+		pctx, r.cancelPull = context.WithCancel(context.Background()) // TODO(stevvooe): Bind a context to the entire controller.
+
+		go func() {
+			defer close(r.pulled)
+			r.pullErr = r.adapter.pullImage(pctx)
+		}()
+	}
+
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-r.pulled:
+		if r.pullErr != nil {
+			// NOTE(stevvooe): We always try to pull the image to make sure we have
+			// the most up to date version. This will return an error, but we only
+			// log it. If the image truly doesn't exist, the create below will
+			// error out.
+			//
+			// This gives us some nice behavior where we use up to date versions of
+			// mutable tags, but will still run if the old image is available but a
+			// registry is down.
+			//
+			// If you don't want this behavior, lock down your image to an
+			// immutable tag or digest.
+			log.G(ctx).WithError(r.pullErr).Error("pulling image failed")
+		}
+	}
+
+	if err := r.adapter.create(ctx); err != nil {
```

Comment:
IMHO it'd be interesting/useful to try, assuming we want to continue down the route of mashing `ContainerSpec` in to containerd as opposed to adding `ContainerdSpec` (gah, naming).